MM-27498 Implement new unread filter

hmhealey opened the pull request:
This PR implements the new version of the unread filter as specced out [here](https://www.figma.com/proto/zVNYaShAXDK8i1RdpuEkr8/Channel-Sidebar---All-Phases?node-id=2185%3A1327&viewport=-5533%2C4423%2C1&scaling=min-zoom). Instead of the unread filter behaving the same as temporarily collapsing all of the channel categories, it now condenses all channels into a single category similar to the old Unreads category that existed in prior to the new sidebar.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-27498

#### Screenshots
![Screen Shot 2020-09-30 at 3 33 43 PM](https://user-images.githubusercontent.com/3277310/94731188-5be64d80-0332-11eb-8f95-e7edacc4babd.png)
![Screen Shot 2020-09-30 at 3 33 51 PM](https://user-images.githubusercontent.com/3277310/94731191-5db01100-0332-11eb-804c-88c300bf32eb.png)

hmhealey commented on components/sidebar/sidebar_category_header.tsx:
```diff
@@ -0,0 +1,56 @@
+    onClick?: (event: React.MouseEvent<HTMLElement>) => void;
+}
+
+const SidebarCategoryHeader = React.forwardRef((props: Props, ref?: React.Ref<HTMLDivElement>) => {
```
I pulled this out of the `SidebarCategory` into a reusable component so that I didn't need to repeat the styles in the `UnreadChannels` component

hmhealey commented on components/sidebar/sidebar_category_header.tsx:
```diff
@@ -0,0 +1,56 @@
+const SidebarCategoryHeader = React.forwardRef((props: Props, ref?: React.Ref<HTMLDivElement>) => {
+    return (
+        <div className='SidebarChannelGroupHeader'>
+            <div
```
This has changed from a button to a clickable div. I hoped to switch between the two components based on whether or not the header should be clickable, but I ran into typing issues because of the ref. This seems to work well enough though since the header can still be tabbed through as before.

hmhealey commented on components/sidebar/sidebar_channel/sidebar_channel.tsx:
```diff
@@ -128,40 +140,63 @@ export default class SidebarChannel extends React.PureComponent<Props, State> {
+        let wrappedComponent: React.ReactNode;
+
+        if (isDraggable) {
+            wrappedComponent = (
```
It might be nice if we had some sort of `ConditionalDraggable` component that could toggle this behaviour via a prop, but I wasn't able to do that at this time.

hmhealey commented on components/sidebar/sidebar_channel_list/sidebar_channel_list.tsx:
```diff
@@ -410,7 +409,44 @@ export default class SidebarCategoryList extends React.PureComponent<Props, Stat
-        const renderedCategories = categories.map(this.renderCategory);
+
+        let channelList: React.ReactNode;
+        if (this.props.isUnreadFilterEnabled) {
```
Swapping `UnreadChannels` for a set of `ChannelCategory` components gives us things like the `More Unreads` indicators for free as long as we still set the refs for each channel correctly.

hmhealey commented on components/sidebar/unread_channels/index.ts:
```diff
@@ -0,0 +1,18 @@
+
+function mapStateToProps(state: GlobalState) {
+    return {
+        unreadChannels: getUnreadChannels(state),
```
This could theoretically be passed from the `SidebarChannelList`, but I decided to call the selector again for the sake of clarity.

hmhealey commented on sass/layout/_sidebar-left.scss:
```diff
@@ -599,26 +599,24 @@
-            //background: radial-gradient(60.83% 3421.87% at 50% 50%, rgba(255, 255, 255, 0.4) 0%, rgba(255, 255, 255, 0) 100%);
-            box-shadow: inset -1px 0 0 rgba(var(--sidebar-text-rgb), 0.60), inset 0 -1px 0 rgba(var(--sidebar-text-rgb), 0.6), inset 1px 0 0 rgba(var(--sidebar-text-rgb), 0.6), inset 0 1px 0 rgba(var(--sidebar-text-rgb), 0.6);
-            border-radius: 4px;
+    .SidebarChannelGroup:not(.dmCategory) .draggingOver .SidebarChannelGroupHeader:hover {
```
The previous code applied the `draggingOverHeader` class only sometimes depending on the type of the category. Since the UNREAD CHANNELS header doesn't have a category, I chose to always apply the class but add an additional class to the category to identify the DMs and let the CSS determine what styles DMs need vs other categories.

hmhealey commented on selectors/views/channel_sidebar.ts:
```diff
@@ -76,6 +74,67 @@ export function makeGetCurrentlyDisplayedChannelsForTeam(): (state: GlobalState,
+
+// getUnreadChannels returns an array of all unread channels on the current team for display with the unread filter
+// enabled. Channels are sorted by recency with channels containing a mention grouped first.
+export const getUnreadChannels = (() => {
```
Some of this logic could have been reused from the old sidebar, but that has a lot of cruft in it, so I chose to rewrite what I could. In particular, the sorting logic is copied from the old sidebar, but this one has a lot fewer dependencies than that one does, and it doesn't unnecessarily rewrite channel objects for no reason.

hmhealey commented on selectors/views/channel_sidebar.ts:
```diff
@@ -1,60 +1,61 @@
-export function makeGetCollapsedStateForAllCategoriesByTeam(): (state: GlobalState, teamId: string) => Record<string, boolean> {
+// getChannelsInCategoryOrder returns an array of channels on the current team that are currently visible in the sidebar.
+// Channels are returned in the same order as in the sidebar.
+export const getChannelsInCategoryOrder = (() => {
```
I realize that previously passing in the team ID as a parameter might have been good for letting us reuse selectors, but it also made them much more complicated since you had to worry about making a selector factory to avoid mucking up memoization.

esethna reviewed (approved):
@hmhealey looks good, only tweak would be the text should say "All Unreads" when the filter is enabled

![image](https://user-images.githubusercontent.com/14333569/94738813-d14b2180-0324-11eb-92ae-2c5740290128.png)

hmhealey commented:
As I just found, there's a potential bug here about channels moving when you click on them and they become current. I need to go see how the old Unreads category behaved since it will likely be something similar

esethna commented:
@andrewbrown00 @hmhealey how are muted channels with unreads or mentions expected to sort here?

hmhealey commented:
I believe I made them sort like the Unread category previously did where muted channels are treated the same as unmentioned ones for sorting

esethna commented:
@andrewbrown00 thoughts on this? I'm wondering if we want to surface muted channels at all in the unread filter? Or alternatively at the bottom of the list of unreads when the filter is enabled. Thoughts?

andrewbrown00 commented:
> @andrewbrown00 thoughts on this? I'm wondering if we want to surface muted channels at all in the unread filter? Or alternatively at the bottom of the list of unreads when the filter is enabled. Thoughts?

@esethna @hmhealey let's move muted channels with user mentions/direct messages below all unmuted channels/direct messages when the filter is enabled.

hmhealey commented:
@andrewbrown00 @esethna I've updated it to sort muted channel mentions beneath other unread channels now, and I've fixed the bug I found in R&D where switching to a channel with a mention while the unread filter was enabled would cause it to move in the RHS.
<img width="254" alt="Screen Shot 2020-10-07 at 5 42 41 PM" src="https://user-images.githubusercontent.com/3277310/95391593-52765b80-08c5-11eb-8340-ea2221fee176.png">

@deanwhillier @devinbinnie  I've fixed the issue where we weren't using semantic markup for the header button. I've also made some other changes as mentioned above, so it's ready for a re-review now.

hmhealey commented on types/store/index.ts:
```diff
@@ -47,7 +47,7 @@ export type GlobalState = BaseGlobalState & {
             focusedPostId: string;
             mobileView: boolean;
-            keepChannelIdAsUnread: Channel | null;
+            lastUnreadChannel: (Channel & {hadMentions: boolean}) | null; // Actually only an object with {id: string, hadMentions: boolean}
```
I forget why the name of this field in the store was different than what we called it literally everywhere else, so I renamed it.

The ugly type is because I didn't want to go change this all throughout the channel selectors in mattermost-redux where it's used. We can do that cleanup once the old sidebar is removed and most of those selectors are removed.

hmhealey commented on selectors/views/channel_sidebar.ts:
```diff
@@ -76,6 +74,84 @@ export function makeGetCurrentlyDisplayedChannelsForTeam(): (state: GlobalState,
+            }
+
+            function hasMentions(channel: Channel) {
+                if (lastUnreadChannel && channel.id === lastUnreadChannel.id && lastUnreadChannel.hadMentions) {
```
This `lastUnreadChannel.hadMentions` logic keeps track of the current channel and whether or not it had mentions when we switched to it, and we can use that to sort the current channel as if it did have mentions. It was part of a previous attempt for this exact same thing with the Unreads section of the old sidebar (thanks, past me!), but it mysteriously isn't used in the current iteration of that code, so it took some time to find.

andrewbrown00 reviewed (approved):
Looks good @hmhealey thanks!

![image](https://user-images.githubusercontent.com/48559658/95889631-a1eada80-0d50-11eb-90d6-31a0a877613c.png)

amyblais commented:
@devinbinnie @deanwhillier Kind reminder to review this PR Eric wanted to aim this for v5.29.

deanwhillier commented:
Thanks @amyblais. Was talking with Harrison about this earlier today. There is one more update needed based on our conversation and then I'll do my final.

hmhealey commented:
@devinbinnie @deanwhillier This should be good to review now

deanwhillier reviewed (commented):
Quick comment: looks like the unread header needs a bit of CSS to compensate for the space the expand/collapse arrow was taking up in the expandable header.

![image](https://user-images.githubusercontent.com/3245614/96596179-702ac400-12ba-11eb-8120-fc606c920333.png)

deanwhillier commented on components/sidebar/sidebar_category_header.tsx:
```diff
@@ -0,0 +1,78 @@
+                className={classNames('SidebarChannelGroupHeader_groupButton', {
+                    dragging: props.isDragging,
+                })}
+                aria-label={props.displayName}
```
This isn't needed on the static header.

deanwhillier commented on components/sidebar/sidebar_category_header.tsx:
```diff
@@ -0,0 +1,78 @@
+                })}
+                aria-label={props.displayName}
+            >
+                <div {...props.dragHandleProps}>
+                    {wrapEmojis(props.displayName)}
+                </div>
```
Do we need the `dragHandleProps` code here given it's a static header?

hmhealey replied:
Oh static is a much better name for it. I would've removed those both from the static one, but I couldn't come up with a better name than something like `SidebarCategoryHeaderNotCollapsibleOrDraggable` 😛 . Static vs not static is much better though since it also lets the collapsible/draggable one be the primary one

deanwhillier commented on components/sidebar/sidebar_category_header.tsx:
```diff
@@ -0,0 +1,78 @@
+            <div
+                ref={ref}
+                className={classNames('SidebarChannelGroupHeader_groupButton', {
+                    dragging: props.isDragging,
```
Do we need the `dragging` class code here given it's a static header?

deanwhillier commented on sass/layout/_sidebar-left.scss:
```diff
@@ -599,26 +599,24 @@
         font-weight: 600;
         text-transform: uppercase;
-        overflow: hidden;
         text-overflow: ellipsis;
```
If you are removing `overflow: hidden;` here, the corresponding `text-overflow: ellipsis` style will have no effect anymore so you should probably remove it as well. Neither appears to be necessary for truncating text here anyways as the child div css appears to be handling it.

```suggestion
```

deanwhillier reviewed (changes requested):
A few items to consider below.

Also, I noticed that the 'NEW' label seems to have decided it doesn't want to align properly anymore 😛:

![image](https://user-images.githubusercontent.com/3245614/96604136-c1d74c80-12c2-11eb-8ffb-f0189f7e47ce.png)

deanwhillier reviewed (approved):
👍

esethna commented:
@jgilliam17 this one is also a priority for v5.29 :)

jgilliam17 commented:
@hmhealey Can you please clarify if Muted unread channels are also supposed to surface on the LHS when filter for All Unreads is enabled, or only Muted channels with mentions? Thanks

hmhealey commented:
Muted channels are only marked as unread when they have a mention, so you can't actually have a muted channel without a mention.

esethna commented:
@jgilliam17 just muted channels with mentions should appear when filtered.

jgilliam17 commented on e2e/cypress/integration/channel_sidebar/unread_filter_spec.js:
```diff
@@ -0,0 +1,170 @@
+        });
+    });
+
+    it('should change the filter label when the unread filter changes state', () => {
```
```suggestion
    it('MM-T3441 should change the filter label when the unread filter changes state', () => {
```

jgilliam17 commented on e2e/cypress/integration/channel_sidebar/unread_filter_spec.js:
```diff
@@ -0,0 +1,170 @@
+        cy.get('.SidebarFilters:contains(All channels)').should('be.visible');
+    });
+
+    it('should not persist the state of the unread filter on reload', () => {
```
```suggestion
    it('MM-T3442 should not persist the state of the unread filter on reload', () => {
```

jgilliam17 commented on e2e/cypress/integration/channel_sidebar/unread_filter_spec.js:
```diff
@@ -0,0 +1,170 @@
+        cy.get('.SidebarChannelGroupHeader:contains(DIRECT MESSAGES)').should('be.visible');
+    });
+
+    it('should only show unread channels with filter enabled', () => {
```
```suggestion
    it('MM-T3443 should only show unread channels with filter enabled', () => {
```

jgilliam17 commented on e2e/cypress/integration/channel_sidebar/unread_filter_spec.js:
```diff
@@ -0,0 +1,170 @@
+        cy.get(`#sidebarItem_${readChannelName}`).should('be.visible').should(beRead);
+    });
+
+    it('should always show the current channel, even if it is not unread', () => {
```
```suggestion
    it('MM-T3444 should always show the current channel, even if it is not unread', () => {
```

jgilliam17 commented on e2e/cypress/integration/channel_sidebar/unread_filter_spec.js:
```diff
@@ -0,0 +1,170 @@
+        disableUnreadFilter();
+    });
+
+    it('should hide channels once they have been read', () => {
```
```suggestion
    it('MM-T3445 should hide channels once they have been read', () => {
```

jgilliam17 reviewed (commented):
Thank you @hmhealey 
Tested, looks good to merge.
-Verified, when enabled, new filter by unread now compress all unread channels into a single category All Unreads
- Verified Unread filter disables on refresh
- Verified muted unread channels with mentions are sorted at the bottom of All Unreads
- Than you for adding E2Es 🎉 I added TM4J test cases and updated with steps, details and links. Could you add one more for muted unread channel with mention?

mattermod commented:
Cherry pick is scheduled.
